File: combina77.java
import java.util.*;
public class combina77 {
    public List<List<Integer>> combine(int n, int k) {
        int[] nums = new int[n];
        for(int i=0;i<n;i++){
            nums[i]=i+1;
        }
        List<Integer> list = new ArrayList<Integer>();
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        backtrack(nums,k,list,result);
        return result;
    }
    public static void backtrack(int[] nums,int k,List<Integer> list,List<List<Integer>> result){
        if(list.size()==k){
            result.add(list);
        }
        else{
            int tmpindex = 0;
            if(list.size()!=0)tmpindex=list.get(list.size()-1);
            for(int i=tmpindex;i<nums.length;i++){
                list.add(nums[i]);
                List<Integer> listtmp = new ArrayList<Integer>();
                listtmp.addAll(list);
                backtrack(nums,k,listtmp,result);
                list.remove(list.size()-1);
            }
        }
    }
}

File: combina77.py
#!/usr/bin/env python
# coding=utf-8
def backtrack(nums,k,list,result):
    if len(list)==k:
        result.add(list)
    else:
        tmpindex = 0
        if len(list)!=0:
            tmpindex=list[len(list)-1]
            for i in range(tmpindex,len(nums)):
                list.append(nums[i])
                listtmp = [n for n in list]
                backtrack(nums,k,list,result)
                list=list[0:-1]
def combine(self,n,k):
    nums = [i+1 for i in range(n)]
    list = []
    result = []
    backtrack(nums,k,list,result)
    return result
